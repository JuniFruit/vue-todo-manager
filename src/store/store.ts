import { IPersonInfo } from "@/components/person-cards/Cards.interface";
import { IProjectItem } from "@/components/projects-list/List.interface";
import Vue from "vue";
export interface IState {
  toastMsg: string;
  toastType: ToastType;
  projects: IProjectItem[];
  users: IPersonInfo[];
}

type ToastType = "success" | "error";

const state: IState = {
  toastMsg: "",
  toastType: "success",
  projects: [],
  users: [],
};

export const store = Vue.observable<IState>({
  ...state,
});

export const mutations = {
  setToast(msg: string, type: ToastType = "success") {
    store.toastMsg = msg;
    store.toastType = type;
  },
  clearToast() {
    store.toastMsg = "";
  },
  setProjects(projects: IProjectItem[]) {
    store.projects = [...projects];
  },
  setUsers(users: IPersonInfo[]) {
    store.users = [...users];
  },
};
