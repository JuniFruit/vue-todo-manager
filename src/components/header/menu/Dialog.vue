<template>
    <v-dialog width="600px" v-model="isOpen">
        <template v-slot:activator="{ on, attrs }">
            <v-btn color="light-blue accent-3 text-button depressed " dark v-bind="attrs" v-on="on">
                Create a task
            </v-btn>
        </template>

        <v-card>
            <v-card-title class="text-h5 white--text  primary ">
                Add a new project
            </v-card-title>

            <v-card-text class="mt-5">
                <v-form class="px-3" @submit.prevent="submit" ref="form">
                    <v-text-field name="Title" label="Title" v-model="title" :rules="inputRules"
                        prepend-icon="mdi-account"></v-text-field>
                    <v-textarea name="Information" label="Info" v-model="info" :rules="inputRules"
                        prepend-icon="mdi-pencil"></v-textarea>
                    <v-menu offset-y min-width="auto">
                        <template v-slot:activator="{ on, attrs }">
                            <v-text-field v-on="on" v-bind="attrs" :rules="inputRules" name="Due" :value="fomatted"
                                prepend-icon="mdi-calendar" label="Due date"></v-text-field>
                        </template>
                        <v-date-picker v-model="due" color="primary lighten-2"></v-date-picker>
                    </v-menu>
                    <v-flex row justify-center>
                        <v-btn :loading="isLoading" type="submit" depressed color="primary my-5">Add a project</v-btn>

                    </v-flex>

                </v-form>
            </v-card-text>


        </v-card>
    </v-dialog>
</template>

<script lang="ts">
//@ts-nocheck
import Vue from 'vue';
import { IProjectItem } from '@/components/projects-list/List.interface';
import { Firestore } from '@/services/firestore.service';
import { mutations } from '@/store/store';
import { format } from 'date-fns';
export default Vue.extend({
    name: 'Dialog',
    data() {
        return {
            title: '',
            info: '',
            due: '',
            inputRules: [
                (v: any) => v?.length >= 3 || 'Minimum length is 3 characters'
            ],
            isLoading: false,
            isOpen: false,
            error: ''
        }
    },
    computed: {
        fomatted() {
            if (!this.due) return;
            const [y, m, d] = this.due.split('-').filter(item => item !== '-')
            return this.due ? format(new Date(+y, +m - 1, +d), "do MMM yyyy") : ''
        }
    },
    methods: {
        async submit() {
            //@ts-ignore
            if (this.$refs.form!.validate()) {
                this.isLoading = true;
                this.error = '';
                const [y, m, d] = this.due.split('-').filter(item => item !== '-')
                const project: IProjectItem = {
                    title: this.title,
                    person: 'User',
                    due: format(new Date(+y, +m - 1, +d), "do MMM yyyy"),
                    content: this.info,
                    status: 'ongoing',
                    id: '0'
                }
                const { isError, message } = await Firestore.addProject(project)
                if (isError) {
                    mutations.setToast(message, 'error')
                } else {
                    mutations.setToast('Successfully added!');
                    this.isOpen = false
                }
                this.isLoading = false

            }
        }
    }

})

</script>