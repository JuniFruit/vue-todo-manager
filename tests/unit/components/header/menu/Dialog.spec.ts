import Dialog from "@/components/header/menu/Dialog.vue";
import { Wrapper } from "@vue/test-utils";
import { createCustomWrapper } from "../../../utils/general";
import { Firestore } from "@/services/firestore.service";
import { mutations } from "@/store/store";

describe("Dialog.vue component", () => {
  let wrapper: Wrapper<Dialog>;
  const TITLE = "Testing project";
  const INFO = "To test fields";
  const DUE = "2023-09-09";
  const ERR_MSG = "Minimum length is 3 characters";
  const SUCC_MSG = "Successfully added!";
  const NET_ERR = "Network Error";
  const mockedPromise = Promise.resolve({ isError: false, message: "" });
  const mockedRejPromise = { isError: true, message: NET_ERR };

  it("renders fields that match snapshot", () => {
    wrapper = createCustomWrapper<Dialog>(Dialog, {
      data() {
        return {
          isOpen: true,
        };
      },
    });
    expect(wrapper.find(".v-dialog__content").exists()).toBe(true);
  });

  it("shows error message when blank from submitted ", async () => {
    wrapper = createCustomWrapper<Dialog>(Dialog, {
      data() {
        return {
          isOpen: true,
        };
      },
    });
    await wrapper.find(".v-form").trigger("submit.prevent");
    expect(wrapper.find(".v-messages__message").text()).toMatch(ERR_MSG);
  });

  describe("when submitted with user inputs", () => {
    let resolvedSpy: any;

    beforeEach(() => {
      resolvedSpy = jest
        .spyOn(Firestore, "addProject")
        .mockImplementation(jest.fn(() => mockedPromise));

      wrapper = createCustomWrapper<Dialog>(Dialog, {
        data() {
          return {
            isOpen: true,
            due: DUE,
          };
        },
      });
    });

    afterEach(() => {
      jest.restoreAllMocks();
      jest.clearAllMocks();
      jest.resetAllMocks();
      wrapper.vm.$destroy();
    });

    it("invokes Firestore service when submitted with project info object", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);

      await wrapper.find(".v-form").trigger("submit.prevent");
      expect(resolvedSpy).toHaveBeenCalledWith({
        title: TITLE,
        person: "User",
        due: "9th Sep 2023",
        content: INFO,
        status: "ongoing",
        id: "0",
      });
    });

    it("sets loading state when submitted", () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);

      wrapper.find(".v-form").trigger("submit.prevent");
      expect(wrapper.vm.$data.isLoading).toBe(true);
    });
    it("sets store toast to success value when promise resolved", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);
      const toast = jest.spyOn(mutations, "setToast");

      await wrapper.find(".v-form").trigger("submit.prevent");

      expect(toast).toHaveBeenCalledWith(SUCC_MSG);
    });

    it("sets store toast to error value when promise rejected", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);
      const toast = jest.spyOn(mutations, "setToast");
      jest
        .spyOn(Firestore, "addProject")
        .mockImplementation(jest.fn(async () => mockedRejPromise));
      await wrapper.find(".v-form").trigger("submit.prevent");

      expect(toast).toHaveBeenCalledWith(NET_ERR, "error");
    });

    it("closes dialog on success", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);

      await wrapper.find(".v-form").trigger("submit.prevent");
      const dialog = wrapper.find(".v-dialog__content--active");
      expect(dialog.exists()).toBe(false);
    });

    it("renders dialog on error", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);
      jest
        .spyOn(Firestore, "addProject")
        .mockImplementation(jest.fn(async () => mockedRejPromise));
      const dialog = wrapper.find(".v-dialog__content--active");
      expect(dialog).toMatchSnapshot();
    });

    it("changes loading state to false on error", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);
      jest
        .spyOn(Firestore, "addProject")
        .mockImplementation(jest.fn(async () => mockedRejPromise));

      expect(wrapper.vm.$data.isLoading).toBe(false);
    });

    it("changes loading state to false on success", async () => {
      wrapper.find("input[name=Title]").setValue(TITLE);
      wrapper.find("textarea[name=Information]").setValue(INFO);

      expect(wrapper.vm.$data.isLoading).toBe(false);
    });
  });
});
