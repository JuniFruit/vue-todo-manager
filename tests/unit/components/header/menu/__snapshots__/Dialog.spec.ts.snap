// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog.vue component when submitted with user inputs renders dialog on error 1`] = `
<div role="dialog" aria-modal="true" class="v-dialog__content v-dialog__content--active" style="z-index: 202;">
  <transition-stub name="dialog-transition" appear="true">
    <div tabindex="0" class="v-dialog v-dialog--active" style="transform-origin: center center; width: 600px;">
      <div class="v-card v-sheet theme--light">
        <div class="v-card__title text-h5 white--text primary">
          Add a new project
        </div>
        <div class="v-card__text mt-5">
          <form novalidate="novalidate" class="v-form px-3">
            <div class="v-input theme--light v-text-field">
              <div class="v-input__prepend-outer">
                <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-account theme--light"></i></div>
              </div>
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-174" class="v-label theme--light" style="left: 0px; position: absolute;">Title</label><input name="Title" id="input-174" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-input v-textarea theme--light v-text-field">
              <div class="v-input__prepend-outer">
                <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-pencil theme--light"></i></div>
              </div>
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-178" class="v-label theme--light" style="left: 0px; position: absolute;">Info</label><textarea name="Information" id="input-178" rows="5"></textarea></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="v-menu">
              <!---->
            </div>
            <div class="v-input v-input--is-label-active v-input--is-dirty theme--light v-text-field">
              <div class="v-input__prepend-outer">
                <div class="v-input__icon v-input__icon--prepend"><i aria-hidden="true" class="v-icon notranslate mdi mdi-calendar theme--light"></i></div>
              </div>
              <div class="v-input__control">
                <div class="v-input__slot">
                  <div class="v-text-field__slot"><label for="input-183" class="v-label v-label--active theme--light" style="left: 0px; position: absolute;">Due date</label><input name="Due" role="button" aria-haspopup="true" aria-expanded="false" id="input-183" type="text"></div>
                </div>
                <div class="v-text-field__details">
                  <div class="v-messages theme--light">
                    <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
                  </div>
                </div>
              </div>
            </div>
            <div class="flex row justify-center"><button type="submit" class="v-btn v-btn--has-bg theme--light v-size--default primary my-5"><span class="v-btn__content">Add a project</span></button></div>
          </form>
        </div>
      </div>
    </div>
  </transition-stub>
</div>
`;
